crates/holochain_types/src/chain.rs

```diff
@@ -20,3 +28,155 @@ pub trait AgentActivityExt {
 }
 
 impl AgentActivityExt for AgentActivityResponse {}
+
+#[warn(missing_docs)]
+#[must_use = "Iterator doesn't do anything unless consumed."]
+/// Iterate over a source chain and apply the [`ChainFilter`] to each element.
+/// This iterator will:
+/// - Ignore any ops that are not a direct ancestor to the starting position.
+/// - Stop at the first gap in the chain.
+/// - Take no **more** then the [`take`]. It may return less.
+/// - Stop at (including) the [`ActionHash`](holo_hash::ActionHash) in [`until`]. But not if this hash is not in the chain.
+///
+/// [`take`]: ChainFilter::take
+/// [`until`]: ChainFilter::until
+pub struct ChainFilterIter<I>
+where
+    I: Iterator<Item = RegisterAgentActivityOp>,
+{
+    filter: ChainFilter,
+    iter: Peekable<I>,
+    end: bool,
+}
+
+#[warn(missing_docs)]
+impl<I> ChainFilterIter<I>
+where
+    I: Iterator<Item = RegisterAgentActivityOp>,
+{
+    /// Create an iterator that filters an iterator of [`RegisterAgentActivityOp`]
+    /// with a [`ChainFilter`].
+    ///
+    /// # Constraints
+    /// - The input iterator **must** be sorted by action sequence from highest to lowest.
+    /// - If the input iterator is not sorted by action sequence
+    /// then this iterator will only work on the first sorted subset.
+    /// - If the iterator does not contain the filters starting position
+    /// then this will be an empty iterator.
+    pub fn new(
+        filter: ChainFilter,
+        iter: impl IntoIterator<Item = RegisterAgentActivityOp, IntoIter = I>,
+    ) -> Self {
+        // Create a peekable iterator.
+        let mut iter = iter.into_iter().peekable();
+
+        // Discard any ops that are not the starting position.
+        let i = iter.by_ref();
+        while let Some(op) = i.peek() {
+            if *op.action.action_address() == filter.position {
+                break;
+            }
+            i.next();
+        }
+
+        Self {
+            filter,
+            iter,
+            end: false,
+        }
+    }
+}
+
+impl<I> Iterator for ChainFilterIter<I>
+where
+    I: Iterator<Item = RegisterAgentActivityOp>,
+{
+    type Item = RegisterAgentActivityOp;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.end {
+            return None;
+        }
+
+        let op = self.iter.next()?;
+        let op = loop {
+            let parent = self.iter.peek();
+
+            // Check the next sequence number
+            match parent {
+                Some(parent) => {
+                    let child_seq = op.action.hashed.action_seq();
+                    let parent_seq = parent.action.hashed.action_seq();
+                    match (child_seq.cmp(&parent_seq), op.action.hashed.prev_action()) {
+                        (std::cmp::Ordering::Less, _) => {
+                            // The chain is out of order so we must end here.
+                            self.end = true;
+                            break op;
+                        }
```

Comment:
So I have updated this in the follow up PR #1483 so that it does the sorting as I tested it and it's actually very cheap to sort in memory instead of the database. I will try and pull those changes into this PR if I can figure out the git foo needed.

Reply:
Ok this should be resolved now